### PR TITLE
CWL document resolver via GA4GH tool API 

### PR DESCRIPTION
**Pull request by denis-yuen**
This allows cwltool to launch tools from https://dockstore.org/ in one of two ways

```
cwltool --non-strict quay.io/collaboratory/dockstore-tool-bamstats:master test.json
```

and

```
# defaults to latest when a version is not specified
cwltool --non-strict quay.io/collaboratory/dockstore-tool-bamstats test.json
```

Grab the test.json (and input file) from https://github.com/CancerCollaboratory/dockstore-tool-bamstats

### CODE REVIEW

**Comment by denis-yuen** [comment]
@tetron FYI, looks like something has gone wrong stylistically with the merge from master

**Comment by mr-c** [comment]
@denis-yuen The conformance tests are failing: https://ci.commonwl.org/job/cwltool-pr-conformance/275/console

**Comment by denis-yuen** [comment]
@mr-c Yeah, just got travis satisfied. Looks like Jenkins is next if I can figure this out.

**Comment by mr-c** [comment]
Looks useful to me, shall I update the branch and merge?

**Comment by mr-c** [comment]
@denis-yuen FYI, I think you have permissions to create branches on this repo directly

**Comment by denis-yuen** [comment]
@mr-c Ah ok, thanks. 
Feel free to update and merge/squash

**Comment by tetron** [comment]
Hi @denis-yuen can you update this PR?

**Comment by denis-yuen** [comment]
@tetron FYI, done

**Comment by denis-yuen** [comment]
@tetron  @mr-c Hi, not sure what is the next step with this. Just give me a heads-up if more is needed on my side.

**Comment by mr-c** [comment]
Hey @denis-yuen thank you for this contribution!

The documentation at the top should be added to the readme as a general section about resolving the location of referenced documents, with a link to http://www.commonwl.org/v1.0/CommandLineTool.html#Discovering_CWL_documents_on_a_local_filesystem

Specifically it must be clearly explained that the reference implementation will try to contact dockstore.org if the document can't be found locally; likewise there should be a switch to skip dockstore.org resolution for privacy preservation.

- [ ] docs
- [ ] opt-out switch

**Comment by denis-yuen** [comment]
FYI @tetron , I may need you (or someone more pythony) to finish off this since my Python knowledge is lacking. 
It looks like the immediate issue is related to this https://github.com/python/mypy/issues/1141 . In addition, there's probably a more elegant of passing the opt-out parameter.

**Comment by mr-c** [comment]
@denis-yuen I'll fix the types for you

**Comment by mr-c** [comment]
@denis-yuen A few questions:

Can you help me understand the value in a dockstore.org specific resolver when we already support any HTTP[S] IRI?

What's the benefit of typing:

```
cwltool --non-strict https://raw.githubusercontent.com/CancerCollaboratory/dockstore-tool-bamstats/develop/Dockstore.cwl test.json
```
(which works today, and won't need the `--non-strict` once [my PR](https://github.com/CancerCollaboratory/dockstore-tool-bamstats/pull/4) is merged)
over
```
cwltool --non-strict quay.io/collaboratory/dockstore-tool-bamstats:master test.json
```
?

As an end user I'm not likely to want to type `quay.io/collaboratory/dockstore-tool-bamstats:master` from memory, so if we stay in the land of copy and paste then a regular URL/URI/IRI seems fine.

As a shortcut, the user's input object (which really should be YAML not JSON for readability) can include a `cwl:tool` entry pointing to the remote location of the tool or workflow. So maybe the docstore.org website or tool can generate a skelton input object YAML with that prefilled?

**Comment by denis-yuen** [comment]
@mr-c 
Well, the alternative to `quay.io/collaboratory/dockstore-tool-bamstats:master` would be something more like `https://dockstore.org:8443/api/ga4gh/v1/tools/quay.io/collaboratory/dockstore-tool-bamstats/versions/master/plain-CWL/descriptor`.  `https://raw.githubusercontent.com/CancerCollaboratory/dockstore-tool-bamstats/develop/Dockstore.cwl`is different in the sense that its pointing to the GitHub repo where Dockstore sources its content and is pointing at GitHub's develop rather than dockstore/quay.io's docker tag. 

As for whether copy-and-paste is sufficient, I think @tetron 's intent from the original commit was that this is more meant for ease-of-use, so that someone can go to the dockstore site and copy just `quay.io/collaboratory/dockstore-tool-bamstats:master` rather than the long URL above. 

Oh, I was not aware of that last short-cut. Could you point me to an example in the user guide or similar?

**Comment by mr-c** [comment]
`cwl:tool` is described at http://www.commonwl.org/v1.0/CommandLineTool.html#Executing_CWL_documents_as_scripts

**Comment by mr-c** [comment]
@denis-yuen To keep the code cleaner and more maintainable, and to be fair to other registries, I suggest the following:

- [ ] @mrc adds support for resolver plugins using the `cwltool.resolver` [entry point](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points)
- [ ] @denis-yuen and/or others implement their additional resolvers and advertise them using the `cwltool.resolver` [entry point](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points)
- [ ] Once the above is working and tested, dependencies will be added to `cwltool` as named [extras](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) so that users can install them easier. Example: `pip install cwltool[docstore,bioshadock]`

**Comment by mrc** [comment]
I reckon I've been CC'd here by mistake, which means you might be missing a
review from someone.

**Comment by mr-c** [comment]
@mrc Sorry, typo'd my own username :-)

**Comment by denis-yuen** [comment]
@mr-c FYI, sounds reasonable, but it may take some time before I can get to it.

**Comment by mrc** [comment]
Hehe no worries.

**Comment by tetron** [comment]
@mr-c I don't object to implementing a proper plugin framework.  That would provide a much better solution to configuring the various extension points than the ad-hoc current approach.  However I have limited bandwidth so my personal preference would be to go ahead and merge the feature as-is and refactor later.

Do you have any experience / opinions on plugin frameworks for Python that we can use?

File: README.rst [hunk 1]
```diff
@@ -86,3 +86,16 @@ documents using absolute or relative local filesytem paths. If a relative path
 is referenced and that document isn't found in the current directory then the
 following locations will be searched:
 http://www.commonwl.org/v1.0/CommandLineTool.html#Discovering_CWL_documents_on_a_local_filesystem
+
+Use with Dockstore
```
Comment (mr-c) [inline README.rst]: All references to Dockstore need to be replaced with "the GA4GH Tool Registry API"

At some point there should be a link to https://github.com/ga4gh/tool-registry-schemas#what-is-the-tool-registry-api-schema

Dockstore.org should still be mentioned as it is the default backend

File: cwltool/main.py [hunk 2]
```diff
@@ -162,7 +162,16 @@ def arg_parser():  # type: () -> argparse.ArgumentParser
...
 
-    parser.add_argument("--on-error", type=str,
+    exgroup = parser.add_mutually_exclusive_group()
+    exgroup.add_argument("--enable-tool-registry", action="store_true", help="Enable resolution using tool registry",
+                        dest="enable_tool_registry", default=True)
+    exgroup.add_argument("--disable-tool-registry", action="store_false", help="Disable resolution using registry",
+                        dest="enable_tool_registry", default=True)
+
+    parser.add_argument("--add-tool-registry", action="append", help="Add a tool registry to use for resolution, default %s" % tool_registries,
+                        dest="tool_registries", default=[])
```
Comment (mr-c) [inline cwltool/main.py]: The help text for all three of theses needs to mention what kind of tool registry; in this case it is "a GA4GH compliant tool registry"

See https://github.com/ga4gh/tool-registry-schemas/issues/17 for a discussion on this

**Comment by mr-c** [comment]
@tetron Sure, we can merge as soon as the docs and help for the new command line flags in the PR are brought up to date with the latest changes (which center the GA4GH API and use Dockstore as the exemplar default)

As for plugins, I've expanded my [comment above](https://github.com/common-workflow-language/cwltool/pull/201#issuecomment-259941214) into https://github.com/common-workflow-language/cwltool/issues/314

**Pull request merged**